base/reinterpretarray.jl

```diff
@@ -765,3 +769,86 @@ end
 
 mapreduce_impl(f::F, op::OP, A::AbstractArrayOrBroadcasted, ifirst::SCartesianIndex2, ilast::SCartesianIndex2) where {F,OP} =
     mapreduce_impl(f, op, A, ifirst, ilast, pairwise_blocksize(f, op))
+
+@pure function struct_subpadding(::Type{Out}, ::Type{In}) where {Out, In}
```

Comment:
Shouldn't this use `@assume_effects` instead?

Reply:
Presumably, I saw something about that switchover.

Reply:
`@assume_effects :foldable` I suppose is a given based on the docs, but @andyferris wrote that method originally so I'm not sure why `@pure` was needed, and I don't know what other effect arguments make sense.

Reply:
Yeah that code is just old - it should be upgraded to use `@assume_effects` (if an annotation is necessary at all, these days)

Reply:
What would show if it's needed or not? Compilation time?

Reply:
To be honest, I doubt any annotation is needed at all. The functions are fairly simple.

If the allocation in `padding` prevents some elimination, it should be possible to switch the code using it here to a generator.

Reply:
Turns out it was needed. Without the annotations, was getting 77 allocations an taking over 2 microseconds, vs. 0 allocations and < 0.8 microseconds